packages/babel-parser/src/tokenizer/state.ts

```diff
@@ -152,20 +279,36 @@ export default class State {
   }
 
   clone(skipArrays?: boolean): State {
-    const state = new State();
-    const keys = Object.keys(this) as (keyof State)[];
-    for (let i = 0, length = keys.length; i < length; i++) {
-      const key = keys[i];
-      let val = this[key];
-
-      if (!skipArrays && Array.isArray(val)) {
-        val = val.slice();
-      }
-
-      // @ts-expect-error val must conform to S[key]
-      state[key] = val;
+    function maybeCopy(val: any[]) {
+      return skipArrays ? val : val.slice();
     }
 
+    const state = new State();
+    state.flags = this.flags;
+    state.curLine = this.curLine;
+    state.lineStart = this.lineStart;
+    state.startLoc = this.startLoc;
+    state.endLoc = this.endLoc;
+    state.errors = maybeCopy(this.errors);
+    state.potentialArrowAt = this.potentialArrowAt;
+    state.noArrowAt = maybeCopy(this.noArrowAt);
+    state.noArrowParamsConversionAt = maybeCopy(this.noArrowParamsConversionAt);
+    state.topicContext = this.topicContext;
+    state.labels = maybeCopy(this.labels);
+    state.commentsLen = this.commentsLen;
+    state.commentStack = maybeCopy(this.commentStack);
+    state.pos = this.pos;
+    state.type = this.type;
+    state.value = this.value;
+    state.start = this.start;
+    state.end = this.end;
+    state.lastTokEndLoc = this.lastTokEndLoc;
+    state.lastTokStartLoc = this.lastTokStartLoc;
+    state.context = maybeCopy(this.context);
+    state.firstInvalidTemplateEscapePos = this.firstInvalidTemplateEscapePos;
+    state.strictErrors = this.strictErrors;
+    state.tokensLength = this.tokensLength;
```

Comment:
Do you think we can somehow type-check this list to make sure we don't forget to add props? If not, can you add a comment at the beginning of `State`'s body saying that when adding a new property we must also add it here?